src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockFetcher.scala

```diff
@@ -17,224 +14,215 @@ import io.iohk.ethereum.blockchain.sync.regular.BlockFetcherState.{
 import io.iohk.ethereum.utils.FunctorOps._
-import monix.eval.Task
 import monix.execution.{Scheduler => MonixScheduler}
 import mouse.all._
+import akka.actor.typed.scaladsl.adapter._
 
 import scala.concurrent.duration._
 
 class BlockFetcher(
-    val peersClient: ActorRef,
-    val peerEventBus: ActorRef,
-    val supervisor: ActorRef,
+    val peersClient: ClassicActorRef,
+    val peerEventBus: ClassicActorRef,
+    val supervisor: ClassicActorRef,
     val syncConfig: SyncConfig,
-    val blockValidator: BlockValidator
-) extends Actor
-    with ActorLogging {
+    val blockValidator: BlockValidator,
+    context: ActorContext[BlockFetcher.FetchCommand]
+) extends AbstractBehavior[BlockFetcher.FetchCommand](context) {
 
   import BlockFetcher._
 
-  implicit val ec: MonixScheduler = MonixScheduler(context.dispatcher)
-  implicit val sys: ActorSystem = context.system
+  implicit val ec: MonixScheduler = MonixScheduler(context.executionContext)
   implicit val timeout: Timeout = syncConfig.peerResponseTimeout + 2.second // some margin for actor communication
+  private val log = context.log
 
-  private val queue: SourceQueue[BlockFetcherState] = {
-    val cap = 1000
-    val numberOfElements = 1
-    Source
-      .queue[BlockFetcherState](cap, OverflowStrategy.backpressure)
-      .delay(5.seconds, DelayOverflowStrategy.dropHead)
-      .addAttributes(Attributes.inputBuffer(numberOfElements, numberOfElements))
-      .map { s =>
-        log.debug("Resuming fetching with the latest state")
-        fetchBlocks(s.withResumedFetching)
-      }
-      .toMat(Sink.ignore)(Keep.left)
-      .run()
-  }
-
-  override def receive: Receive = idle()
+  val headersFetcher: ActorRef[HeadersFetcher.HeadersFetcherCommand] =
+    context.spawn(
+      HeadersFetcher(peersClient, syncConfig, context.self),
+      "headers-fetcher"
+    )
+  context.watch(headersFetcher)
 
-  override def postStop(): Unit = {
-    super.postStop()
-    peerEventBus ! Unsubscribe()
-  }
+  val bodiesFetcher: ActorRef[BodiesFetcher.BodiesFetcherCommand] =
+    context.spawn(
+      BodiesFetcher(peersClient, syncConfig, context.self),
+      "bodies-fetcher"
+    )
+  context.watch(bodiesFetcher)
 
-  private def idle(): Receive = handleCommonMessages(None) orElse { case Start(importer, blockNr) =>
-    BlockFetcherState.initial(importer, blockValidator, blockNr) |> fetchBlocks
-    peerEventBus ! Subscribe(
-      MessageClassifier(
-        Set(Codes.NewBlockCode, Codes.NewBlockHashesCode, Codes.BlockHeadersCode),
-        PeerSelector.AllPeers
-      )
+  val stateNodeFetcher: ActorRef[StateNodeFetcher.StateNodeFetcherCommand] =
+    context.spawn(
+      StateNodeFetcher(peersClient, syncConfig, context.self),
+      "state-node-fetcher"
     )
-  }
+  context.watch(stateNodeFetcher)
+
+  private def subscribeAdapter(fetcher: ActorRef[BlockFetcher.AdaptedMessageFromEventBus]) =
+    Behaviors.receiveMessage[PeerEventBusActor.PeerEvent] {
```

Comment:
If you only want to receive the `MessageFromPeer` I think you can modify this to be `Behaviors.receiveMessage[MessageFromPeer]` and get rid of the `case _ => Behaviors.unhandled`

Reply:
same here

Reply:
I just tested it and this one compiles if you specify this one message, not the whole trait

Reply:
nice, will change it then, thanks

Reply:
might need to do a `.narrow`? thinking out loud